### PR TITLE
Add support for BandScale zoom

### PR DESCRIPTION
**Pull request by JCQuintas**
- [ ] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/x-charts/src/context/CartesianProvider/computeValue.ts [hunk 1]
```diff
@@ -131,6 +144,8 @@ export function computeValue(
 
     completeAxis[axis.id] = {
       ...axis,
+      visibleData: axis.data!,
+      visibleDataRange: [0, axis.data?.length ?? 0],
```
Comment (alexfauquette) [inline packages/x-charts/src/context/CartesianProvider/computeValue.ts:156]: Looks like a duplicate. What about keeping only `visibleDataRange` and the component could obtain `visibleData` with `visibleData = data.slice(...visibleDataRange)`

File: packages/x-charts/src/hooks/useAxisEvents.ts [hunk 2]
```diff
@@ -31,11 +31,11 @@ export const useAxisEvents = (disableAxisListener: boolean) => {
       return () => {};
     }
 
-    const getUpdate = (axisConfig: AxisDefaultized, mouseValue: number) => {
-      if (usedXAxis === null) {
+    function getNewAxisState(xAxisId: AxisId, axisConfig: AxisDefaultized, mouseValue: number) {
+      if (xAxisId === null) {
         return null;
       }
```
Comment (alexfauquette) [inline packages/x-charts/src/hooks/useAxisEvents.ts]:
```suggestion
    function getNewAxisState(axisConfig: AxisDefaultized, mouseValue: number) {
```

File: packages/x-charts/src/hooks/useAxisEvents.ts [hunk 3]
```diff
@@ -118,8 +118,8 @@ export const useAxisEvents = (disableAxisListener: boolean) => {
         dispatch({ type: 'exitChart' });
         return;
       }
-      const newStateX = getUpdate(xAxis[usedXAxis], svgPoint.x);
-      const newStateY = getUpdate(yAxis[usedYAxis], svgPoint.y);
+      const newStateX = getNewAxisState(usedXAxis, xAxis[usedXAxis], svgPoint.x);
+      const newStateY = getNewAxisState(usedXAxis, yAxis[usedYAxis], svgPoint.y);
```
Comment (alexfauquette) [inline packages/x-charts/src/hooks/useAxisEvents.ts]:
```suggestion
      const newStateX = getNewAxisState(xAxis[usedXAxis], svgPoint.x);
      const newStateY = getNewAxisState(yAxis[usedYAxis], svgPoint.y);
```

File: packages/x-charts/src/hooks/useAxisEvents.ts (hunk 2 above)
Comment (alexfauquette) [reply packages/x-charts/src/hooks/useAxisEvents.ts]: It's a legacy from the time this function got only the updated value

```js
const getUpdateY = (y: number) => {
  if (usedYAxis === null) {
    return null;
   }
   const { scale: yScale, data: yAxisData } = yAxis[usedYAxis];
```

https://github.com/alexfauquette/mui-x/commit/5146ec735317ee03f085b486766681bcd9f8c193#diff-b58562fd2d606dc619009aa3e4032dbb842b69eb223bd2e4e8255ff7d1529e5a

File: packages/x-charts/src/BarChart/BarPlot.tsx [hunk 4]
```diff
@@ -186,7 +190,7 @@ const useAggregatedData = (): {
   });
 
   return {
-    completedData: data,
+    completedData: data.filter(Boolean) as CompletedBarData[],
```
Comment (alexfauquette) [inline packages/x-charts/src/BarChart/BarPlot.tsx:193]: I don't get what's the purpose of this modification. What is it supposed to filter?

**Comment by alexfauquette** [comment]
I like the idea of having the index as a reference for the ordinal scales. But I'm concerned by the modification of the scale with `scaleBand(filteredData, chartRange)` because it implies that `dataIndex` will not be the corect one. It will be the index of the element in the zoomed context which mean users would have to do some computation to get access to their data

I'm wondering if it would make sens to only update the `range` and not the data.

It could impact the tick (not sure because ordinal scale tends to a one tick per item)

This ide come from this example of the D3 creator:
https://observablehq.com/@d3/zoomable-bar-chart?collection=@d3/d3-zoom

```js
function zoomed(event) {
	// Here it updates the range of the x-axis (not the domain)
	x.range([marginLeft, width - marginRight].map(d => event.transform.applyX(d)));
	
	// Then update the rendered axis and bars
	svg.selectAll(".bars rect").attr("x", d => x(d.letter)).attr("width", x.bandwidth());
	svg.selectAll(".x-axis").call(xAxis);
}
```

----


Other question: did you had a look at how other libraries handle the distinction between continuous and ordinal scales?

File: packages/x-charts/src/BarChart/BarPlot.tsx (hunk 4 above)
Comment (JCQuintas) [reply packages/x-charts/src/BarChart/BarPlot.tsx:193]: The code just above it returns null if one of the values is `NaN`. This is so the animations work. But could be optimised eventually to not even calculate it

```
        if (Number.isNaN(result.x) || Number.isNaN(result.y)) {
          return null;
        }
```

**Comment by JCQuintas** [comment]
> I'm wondering if it would make sens to only update the `range` and not the data.
> 

I hand't thought of that. A quick test seems promising, I still have to figure the math out, and  we will have to fix the axis bleeding out as well.

https://github.com/JCQuintas/mui-x/assets/9891233/e7b3117d-98f1-48ba-98b9-c996e8901ead

**Comment by JCQuintas** [comment]
> Other question: did you had a look at how other libraries handle the distinction between continuous and ordinal scales?

I looked at `Recharts` and `Echarts` and both have very different logic for how they handle "zooming". `Echarts` has actual zoom, but it is very different than ours, I've been trying to match the behaviour they offer, but not the actual code. `Recharts` is a "brush" style zoom, and from what I understood from their code, they calculate everything beforehand and operate on indexes

**Pull request closed without merging**